File: src/main/java/com/habit_tracker/backend/challenge/service/ChallengeService.java
package com.habit_tracker.backend.challenge.service;

import com.habit_tracker.backend.challenge.dto.ChallengeDto;
import com.habit_tracker.backend.challenge.dto.ChallengeLibraryDto;
import com.habit_tracker.backend.challenge.dto.CreateChallengeRequest;
import com.habit_tracker.backend.challenge.dto.InviteUsersRequest;
import com.habit_tracker.backend.challenge.entity.Challenge;

import java.util.List;
import java.util.UUID;

public interface ChallengeService {

  List<ChallengeLibraryDto> getLibrary();

  ChallengeDto createChallenge(UUID currentUserId, CreateChallengeRequest request);

  List<ChallengeDto> getMyChallenges(UUID currentUserId);

  ChallengeDto markProgress(UUID currentUserId, UUID challengeId);

  ChallengeDto completeChallenge(UUID currentUserId, UUID challengeId);

  ChallengeDto getChallenge(UUID challengeId, UUID currentUserId);

  void deleteChallenge(UUID currentUserId, UUID challengeId);

  ChallengeDto inviteUsers(UUID currentUserId, UUID challengeId, InviteUsersRequest request);
}

File: mobile/HabitsTracker/HabitsTracker/challenge/ChallengeService.swift
import Foundation

final class ChallengeService {
    static let shared = ChallengeService()

    func getActiveChallenges(token: String) async throws -> [ChallengeDTO] {
        try await APIClient.shared.request(
            "/challenges/my",
            authToken: token,
            responseType: [ChallengeDTO].self
        )
    }

    func markProgress(challengeId: UUID, token: String) async throws -> ChallengeDTO {
        try await APIClient.shared.request(
            "/challenges/\(challengeId)/progress",
            method: "POST",
            authToken: token,
            responseType: ChallengeDTO.self
        )
    }
    
    func createChallenge(_ requestBody: CreateChallengeRequest, token: String) async throws -> ChallengeDTO {
        try await APIClient.shared.request(
            "/challenges/create",
            method: "POST",
            body: requestBody,
            authToken: token,
            responseType: ChallengeDTO.self
        )
    }
    
    func deleteChallenge(id: UUID, token: String) async throws {
        try await APIClient.shared.requestVoid(
            "/challenges/\(id)",
            method: "DELETE",
            authToken: token
        )
    }
    
    func inviteUsers(challengeId: UUID, ids: [UUID], token: String) async throws {
        let body = InviteUsersRequest(opponentIds: ids)
        try await APIClient.shared.requestVoid(
            "/challenges/\(challengeId)/invite",
            method: "POST",
            body: body,
            authToken: token
        )
    }
    
    func getTemplates(token: String) async throws -> [ChallengeTemplateDTO] {
        try await APIClient.shared.request(
            "/challenges/library",
            method: "GET",
            authToken: token,
            responseType: [ChallengeTemplateDTO].self
        )
    }
    
    func getChallenge(id: UUID, token: String) async throws -> ChallengeDTO {
        try await APIClient.shared.request(
            "/challenges/get/\(id)",
            method: "GET",
            authToken: token,
            responseType: ChallengeDTO.self
        )
    }

}
